### PR TITLE
Test environment for users microservice

### PR DESCRIPTION
**Pull request by foobazdev**
1. Configure jest config for unit tests
2. Configure jest and docker compose for e2e tests
3. Create script for generation .env files
4. Update README.md

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "version": "1.0.0",
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "install:all": "npm i && npm i --prefix users",
+    "install:all": "npm i && (cd users && npm i && npx prisma generate)",
```
Comment (M-Ziusko) [inline backend/package.json:6]: add postinstall script for npx prisma generate

File: backend/users/src/users/users.controller.ts [hunk 2]
```diff
@@ -27,8 +31,13 @@ export class UsersController {
   }
 
   @Get(':id')
-  findOne(@Param('id', ParseUUIDPipe) id: string) {
-    return this.usersService.findOne(id);
+  async findOne(@Param('id', ParseUUIDPipe) id: string) {
+    const result = await this.usersService.findOne(id);
+    if (Object.keys(result).length === 0) {
```
Comment (M-Ziusko) [inline backend/users/src/users/users.controller.ts:36]:
```suggestion
    
    if (!Object.keys(result).length) {
```

File: backend/users/src/users/users.controller.ts [hunk 3]
```diff
@@ -27,8 +31,13 @@ export class UsersController {
...
+      throw new NotFoundException();
+    }
```
Comment (M-Ziusko) [inline backend/users/src/users/users.controller.ts:38]: Should be deleted cause in service we did same job

File: backend/users/src/users/users.service.ts [hunk 4]
```diff
@@ -25,8 +28,14 @@ export class UsersService {
...
   }
 
-  findOne(id: string) {
-    return this.prisma.user.findUniqueOrThrow({ where: { id } });
+  async findOne(id: string) {
+    const result = await this.prisma.user.findUnique({ where: { id } });
+
+    if (!result) {
+      throw new NotFoundException();
+    }
```
Comment (M-Ziusko) [inline backend/users/src/users/users.service.ts:36]: Should be added as well for PUT :id & DELETE :id

File: backend/users/test/truncate-tables.util.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  const tables = tablenames
+    .map(({ tablename }) => tablename)
+    .filter((name) => name !== '_prisma_migrations')
+    .map((name) => `"public"."${name}"`)
+    .join(', ');
+
+  try {
+    await prisma.$executeRawUnsafe(`TRUNCATE TABLE ${tables} CASCADE;`);
+  } catch (error) {
+    console.log({ error });
```
Comment (M-Ziusko) [inline backend/users/test/truncate-tables.util.ts:19]:
```suggestion
    console.log({ truncateTablesError });
```

File: backend/users/test/users.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  it('/users (GET) should return 200', () =>
+    request(app.getHttpServer()).get('/users').expect(200));
+
+  it('/users (GET) should return empty array if there are no users in database', () =>
+    request(app.getHttpServer())
+      .get('/users')
+      .then((response) => {
+        expect(Array.isArray(response.body)).toBeTruthy();
+        expect(response.body).toHaveLength(0);
+      }));
```
Comment (M-Ziusko) [inline backend/users/test/users.e2e-spec.ts:55]: I would love to use async/await here as well

File: test/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+services:
+  testpostgres:
+    image: postgres:16-alpine
+    container_name: test_postgres
+    restart: always 
+    environment:
+      - POSTGRES_USER=${POSTGRES_USER}
+      - POSTGRES_PASSWORD=${POSTGRES_PASSWORD}
+      - POSTGRES_DB=${POSTGRES_DB}
+      - PGPORT=${POSTGRES_PORT}
```
Comment (M-Ziusko) [inline test/docker-compose.yml:10]:
```suggestion
```

**Review by M-Ziusko — CHANGES_REQUESTED** [review]

File: backend/users/src/users/users.service.ts (hunk 4 above)
Comment (foobazdev) [reply backend/users/src/users/users.service.ts:36]: It is works with Prisma interception filter

File: test/docker-compose.yml (hunk 7 above)
Comment (foobazdev) [reply test/docker-compose.yml:10]: It is necessary  to run postgres on specific port, in this case 5433

File: backend/users/test/util/truncate-tables.util.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+async function truncateTables() {
+  const prisma = new PrismaClient();
+
+  const tablenames = await prisma.$queryRaw<
+    Array<{ tablename: string }>
+  >`SELECT tablename FROM pg_tables WHERE schemaname='public'`;
+
+  const tables = tablenames
+    .map(({ tablename }) => tablename)
+    .filter((name) => name !== '_prisma_migrations')
```
Comment (M-Ziusko) [inline backend/users/test/util/truncate-tables.util.ts:12]: Move to query on line 8

File: backend/users/test/util/truncate-tables.util.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    .map((name) => `"public"."${name}"`)
```
Comment (M-Ziusko) [inline backend/users/test/util/truncate-tables.util.ts:13]:
```suggestion
    .map(({ tablename }) => `"public"."${tablename}"`)
```

File: backend/users/src/users/users.service.ts [hunk 10]
```diff
@@ -25,11 +28,17 @@ export class UsersService {
...
+
+    return result;
   }
 
-  update(id: string, data: UpdateUserDto) {
+  async update(id: string, data: UpdateUserDto) {
```
Comment (M-Ziusko) [inline backend/users/src/users/users.service.ts:41]: Do we need async here?

File: backend/users/src/users/users.service.ts (hunk 4 above)
Comment (M-Ziusko) [reply backend/users/src/users/users.service.ts:36]: Discussed offline

File: backend/users/test/users.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    const res = await request(app.getHttpServer())
+      .post('/users')
+      .send(fakeUser);
+    const userId = res.body.id;
+
+    await request(app.getHttpServer())
+      .patch(`/users/${userId}`)
+      .send({ lastName: 'new last name' })
+      .expect(200);
+  });
```
Comment (M-Ziusko) [inline backend/users/test/users.e2e-spec.ts:88]: I think we should make sure that our schanges are applyed

File: backend/users/test/users.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+  it('/users:id (PATCH) should return 400 if user does not exists', async () => {
```
Comment (M-Ziusko) [inline backend/users/test/users.e2e-spec.ts:90]:
```suggestion
  it('/users:id (PATCH) should return 404 if user does not exists', async () => {
```

**Review by M-Ziusko — CHANGES_REQUESTED** [review]

File: backend/users/test/util/truncate-tables.util.ts (hunk 9 above)
Comment (M-Ziusko) [reply backend/users/test/util/truncate-tables.util.ts:13]: Add cashing for tables list

**Review by M-Ziusko — APPROVED** [review]

**Pull request merged**